File: BillSplitter/cpp/src/BillSplitter.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>
#include "BillSplitter.h"
using namespace std;

BillSplitter::BillSplitter (double total, uint people) {
  this->total = this->round(total);
  this->people = people;
  this->person_index = 0;
}

void BillSplitter::calculate() {
  cout << "Dividing total of $" << fixed << setprecision(2) << this->total << " into " << this->people << " pieces:" << endl;

  double share = this->round(this->total / this->people);
  vector<double> share_list;

  for (uint i = 0; i < this->people; i++) {
    share_list.push_back(share);
  }

  double total = this->round(this->sum(share_list));

  while ( !(this->areEqual(total, this->total)) ) {
      if (this->total - total < 0) {
          this->sortUnevenTotal(share_list, -0.01);
      } else {
          this->sortUnevenTotal(share_list, 0.01);
      }
      total = this->round(this->sum(share_list));
  }

  for (uint num = 0; num < this->people; num++) {
    cout << "Person " << num + 1 << " will pay $" << fixed << setprecision(2) << this->round(share_list.at(num)) << endl;
  }
}

void BillSplitter::sortUnevenTotal(vector<double> &share_list, double incr) {
  share_list.at(person_index) += incr;
  this->person_index++;
  if (person_index >= share_list.size()) {
    this->person_index = 0;
  }
}

bool BillSplitter::areEqual(double var1, double var2) {
  if (abs(var1 - var2) < 0.009999) {
    return true;
  } else {
    return false;
  }
}

double BillSplitter::sum(vector<double> arr) {
  double total = 0.0;
  for (uint i = 0; i < arr.size(); i++) {
    total += arr.at(i);
  }
  return total;
}

double BillSplitter::round(double num) {
  double value = (int)(num * 100 + .5);
  return (double)value / 100;
}

File: BillSplitter/javascript/BillSplitter.js
// bill_splitter.js

class BillSplitter {
  constructor(total, people) {
    this.people = people;
    this.total = total;
    this.person_index = 0;
  }

  _sortUnevenTotal(share_list, incr) {
    share_list[this.person_index] += incr;
    this.person_index++;
    if (this.person_index >= share_list.length) {
      this.person_index = 0;
    }
    return share_list;
  }
  round(number) {
    return Math.round(number * 100) / 100;
  }
  fixed(number) {
    return Number.parseFloat(number).toFixed(2);
  }
  sum(array) {
    var total = 0;
    for (var i = 0; i < array.length; i++) {
      total += array[i];
    }
    return total;
  }
  calculate() {
    console.log(`Dividing total of $${this.fixed(this.total)} into ${this.people} pieces:`)

    var share = this.round(this.total / this.people);
    var share_list = [];

    for (var i = 0; i < this.people; i++) {
      share_list.push(share);
    }

    var total = this.round(this.sum(share_list));

    while ( this.round(total) != this.round(this.total) ) {
        if (this.total - total < 0) {
            this._sortUnevenTotal(share_list, -0.01);
        } else {
            this._sortUnevenTotal(share_list, 0.01);
        }
        total = this.round(this.sum(share_list));
    }

    for (var num = 0; num < this.people; num++) {
      console.log(`Person ${num + 1} will pay $${this.fixed(share_list[num])}`)
    }
  }
}

module.exports = BillSplitter;
